src/bootstrap/cache.rs

```diff
@@ -0,0 +1,267 @@
+    }
+}
+
+impl Deref for Interned<PathBuf> {
+    type Target = Path;
+    fn deref(&self) -> &'static Path {
+        let l = INTERNER.paths.lock().unwrap();
+        unsafe { mem::transmute::<&Path, &'static Path>(l.get(*self)) }
+    }
+}
+
+impl AsRef<Path> for Interned<PathBuf> {
+    fn as_ref(&self) -> &'static Path {
+        let l = INTERNER.paths.lock().unwrap();
+        unsafe { mem::transmute::<&Path, &'static Path>(l.get(*self)) }
+    }
+}
+
+impl AsRef<Path> for Interned<String> {
+    fn as_ref(&self) -> &'static Path {
+        let l = INTERNER.strs.lock().unwrap();
+        unsafe { mem::transmute::<&Path, &'static Path>(l.get(*self).as_ref()) }
+    }
+}
+
+impl AsRef<OsStr> for Interned<PathBuf> {
+    fn as_ref(&self) -> &'static OsStr {
+        let l = INTERNER.paths.lock().unwrap();
+        unsafe { mem::transmute::<&OsStr, &'static OsStr>(l.get(*self).as_ref()) }
+    }
+}
+
+impl AsRef<OsStr> for Interned<String> {
+    fn as_ref(&self) -> &'static OsStr {
+        let l = INTERNER.strs.lock().unwrap();
+        unsafe { mem::transmute::<&OsStr, &'static OsStr>(l.get(*self).as_ref()) }
+    }
+}
+
+
+struct TyIntern<T> {
+    items: Vec<T>,
+    set: HashMap<T, Interned<T>>,
+}
+
+impl<T: Hash + Clone + Eq> TyIntern<T> {
+    fn new() -> TyIntern<T> {
+        TyIntern {
+            items: Vec::new(),
+            set: HashMap::new(),
+        }
+    }
+
+    fn intern_borrow<B>(&mut self, item: &B) -> Interned<T>
+    where
+        B: Eq + Hash + ToOwned<Owned=T> + ?Sized,
+        T: Borrow<B>,
+    {
+        if let Some(i) = self.set.get(&item) {
+            return *i;
+        }
+        let item = item.to_owned();
+        let interned =  Interned(self.items.len(), PhantomData::<*const T>);
+        self.set.insert(item.clone(), interned);
+        self.items.push(item);
+        interned
+    }
+
+    fn intern(&mut self, item: T) -> Interned<T> {
+        if let Some(i) = self.set.get(&item) {
+            return *i;
+        }
+        let interned =  Interned(self.items.len(), PhantomData::<*const T>);
+        self.set.insert(item.clone(), interned);
+        self.items.push(item);
+        interned
+    }
+
+    fn get(&self, i: Interned<T>) -> &T {
+        &self.items[i.0]
+    }
+}
+
+pub struct Interner {
+    strs: Mutex<TyIntern<String>>,
+    paths: Mutex<TyIntern<PathBuf>>,
+}
+
+impl Interner {
+    fn new() -> Interner {
+        Interner {
+            strs: Mutex::new(TyIntern::new()),
+            paths: Mutex::new(TyIntern::new()),
+        }
+    }
+
+    pub fn intern_str(&self, s: &str) -> Interned<String> {
+        self.strs.lock().unwrap().intern_borrow(s)
+    }
+    pub fn intern_string(&self, s: String) -> Interned<String> {
+        self.strs.lock().unwrap().intern(s)
+    }
+
+    pub fn intern_path(&self, s: PathBuf) -> Interned<PathBuf> {
+        self.paths.lock().unwrap().intern(s)
+    }
+}
+
+lazy_static! {
+    pub static ref INTERNER: Interner = Interner::new();
```

Comment:
Ah right yeah forgot to remove this comment as I figured out why as I was reading further, This seems ok.